### ISSUE TITLE
Bugsnag script generates many "reference to undefined property" errors which strict checking is enabled

### ISSUE THREAD

**Issue by birtles** [bugsnag/bugsnag-js#367]
### Expected behavior
No additional errors in console when using bugsnag-js.

### Observed behavior
After importing bugsnag-js I get a bunch of errors from the bugsnag script in my console:
```log
ReferenceError: reference to undefined property "config"
ReferenceError: reference to undefined property "autoNotify"
ReferenceError: reference to undefined property "device"
```

### Steps to reproduce
Import the bugsnag-js module from npm, include it, then call `bugsnag(...)`.

### Version
4.7.2.

### Additional information
Digging into the first error I can see there is the line:

```js
this.config = config.mergeDefaults({ ...this.config, ...opts }, this.configSchema)
```

But until this point nothing ensures that `this.config` is initialized, hence the error. Setting `this.config = {}` in the ctor fixes the problem. I haven't dug into the other errors.

The additional errors in the log obscure the errors I am trying to find. When users report errors they often tell me if they see anything unusual in the console--this is sure to confuse them and generate spurious bug reports.

**Comment by bengourley** [comment]
Hi @birtles, thanks for the report.

I can't reproduce this in any of the browsers I have locally. Can you provide more information or a reduced test case? Which browser are you running this in and how are you bundling your code?

The line of code you mentioned gets compiled down to an `Object.assign()` call, which handles `undefined` values silently, so I don't see why that would be causing the log messages you mentioned.

Let me know what you can provide and I'll be glad to help.

Cheers!

**Comment by birtles** [comment]
I dug into this a little bit and it turns out this error only shows up when the `javascript.options.strict` setting in Firefox is true.

This is set to true when debugging extensions with `web-ext` hence why it doesn't show otherwise.

> The line of code you mentioned gets compiled down to an Object.assign() call, which handles undefined values silently, so I don't see why that would be causing the log messages you mentioned.

The error is not related to `Object.assign()` -- it happens when evaluating the arguments to pass to `Object.assign`. That is, when accessing the `config` property on `this` before calling `Object.assign`.

**Comment by bengourley** [comment]
Ah ok. Thanks for figuring that out! I think this is fine then.

If a user has switched `javascript.options.strict` on then I'm sure they will be used to seeing many warnings such as these! Referencing undefined properties, when it happens unintentionally can be a source of errors, but in these cases the possibility of the accessed value being `undefined` is accounted for.

I'm going to close this off.

**Comment by birtles** [comment]
> If a user has switched javascript.options.strict on then I'm sure they will be used to seeing many warnings such as these!

Hmm, I don't think that's not the case. In this particular project bugsnag is the only library that produces these errors. According to my colleague on the JS team, these errors mostly tend to show up on older scripts such as jQuery and if it weren't for that, we'd turn it on by default in Firefox.

Anyway, if I get a chance I'll work on a PR for this but I don't think I'll get to it soon.

**Comment by bengourley** [comment]
I just switched this option on and navigated to the following websites:

- google.com
- duckduckgo.com
- github.com
- reddit.com
- inbox.google.com

I got a screenful of these warnings on every single one…

<img width="1432" alt="image" src="https://user-images.githubusercontent.com/609579/42818224-34831574-89c8-11e8-86a8-58de48f67d7d.png">

Whilst I don't think this is a good argument in favour of sloppy code, I do think it shows that "_(users of javascript.options.strict)_ will be used to seeing many warnings such as these!".

Regarding the correctness of the code, take the `this.config` example…

```js
// stripped back example of the current code
class Client {
  constructor () {
  }
  configure (opts) {
    this.config = { ...this.config, mergeDefaults(opts, defaults) }
  }
}
```

This code supports repeated and incremental configuration of the client:

```js
var client = new Client()
client.configure({ some: 'options' })
client.configure({ someMore: 'options' })
```

We know that `this.config` will be `undefined` the first time `configure()` is called, but that is ok – we know that `Object.assign()` works the way we'd want it to in this case.

So the code works the way we want it to, for which there are unit tests. Now we are in the business of satisfying the interpreter in this strict mode. In order to satisfy it, we would have to add more code, for example as you mentioned in the constructor:

```js
// stripped back example of the current code
class Client {
  constructor () {
    // either explicitly mark it as undefined – I think
    // this would make the warning go away
    this.config = undefined
    // or make it an empty object, which results in equivalent behaviour
    this.config = {}
  }
  configure (opts) {
    this.config = { ...this.config, mergeDefaults(opts, defaults) }
  }
}
```

Now the warning is gone, but we just added an extra line of code for every undefined property access even though the code was already "correct", and was verified as such with tests.

Since we ship bugsnag on people's webpages, we care a lot about size and how that translates into the performance of our customers' products. Every line of code that ends up in our bundle needs to be justified. For me, preventing these warnings – because they have to be enabled by a hidden browser flag – is not sufficient justification.
